src/include/server/mir/scene/text_input_hub.h

```diff
@@ -0,0 +1,169 @@
+
+#include <memory>
+#include <optional>
+#include <string>
+
+namespace mir
+{
+namespace scene
+{
+namespace detail
+{
+struct TextInputStateSerialTag;
+}
+
+typedef IntWrapper<detail::TextInputStateSerialTag> TextInputStateSerial;
+
+enum class TextInputChangeCause
+{
+    input_method,
+    other,
+};
+
+enum class TextInputContentPurpose
+{
+    normal,
+    alpha,
+    digits,
+    number,
+    phone,
+    url,
+    email,
+    name,
+    password,
+    pin,
+    date,
+    time,
+    datetime,
+    terminal,
+};
+
+enum class TextInputContentHint : uint32_t
+{
+    none                = 0,
+    completion          = 1 << 0,
+    spellcheck          = 1 << 1,
+    auto_capitalization = 1 << 2,
+    lowercase           = 1 << 3,
+    uppercase           = 1 << 4,
+    titlecase           = 1 << 5,
+    hidden_text         = 1 << 6,
+    sensitive_data      = 1 << 7,
+    latin               = 1 << 8,
+    multiline           = 1 << 9,
+};
+
+inline auto operator|(TextInputContentHint a, TextInputContentHint b) -> TextInputContentHint
+{
+    return static_cast<TextInputContentHint>(static_cast<uint32_t>(a) | static_cast<uint32_t>(b));
+}
+
+inline auto operator&(TextInputContentHint a, TextInputContentHint b) -> TextInputContentHint
+{
+    return static_cast<TextInputContentHint>(static_cast<uint32_t>(a) & static_cast<uint32_t>(b));
+}
+
+/// TODO: document (see text-input-unstable-v3.xml)
+struct TextInputState
+{
+    std::optional<std::string> surrounding_text;
+    std::optional<int> cursor;
+    std::optional<int> anchor;
+    std::optional<TextInputChangeCause> change_cause;
+    std::optional<TextInputContentHint> content_hint;
+    std::optional<TextInputContentPurpose> content_purpose;
+};
+
+/// TODO: document (see text-input-unstable-v3.xml)
+struct TextInputChange
+{
+    TextInputChange(TextInputStateSerial serial)
+        : serial{serial}
+    {
+    }
+
+    TextInputStateSerial serial;
+    std::optional<std::string> preedit_text;
+    std::optional<int> preedit_cursor_begin;
+    std::optional<int> preedit_cursor_end;
+    std::optional<std::string> commit_text;
+    std::optional<int> delete_before;
+    std::optional<int> delete_after;
+};
+
+/// Gets notifications about changes in clients state (implemented by input methods)
+class TextInputStateObserver
+{
+public:
+    TextInputStateObserver() = default;
+    virtual ~TextInputStateObserver() = default;
+
+    virtual void activated(TextInputStateSerial serial, bool new_input_field, TextInputState const& state) = 0;
+    virtual void deactivated() = 0;
+
+private:
+    TextInputStateObserver(TextInputStateObserver const&) = delete;
+    TextInputStateObserver& operator=(TextInputStateObserver const&) = delete;
+};
+
+/// Gets notifications of new text entered by the user
+class TextInputChangeHandler
+{
+public:
+    TextInputChangeHandler() = default;
+    virtual ~TextInputChangeHandler() = default;
+
+    virtual void change(TextInputChange const& change) = 0;
```

Comment:
Nit: this is an observer method, right? The name should be passive rather than active - `changed` rather than `change`?

Reply:
Done